File: backend/src/main/java/com/supermarket/stockmanagement/service/PurchaseOrderService.java
package com.supermarket.stockmanagement.service;

import com.supermarket.stockmanagement.model.PurchaseOrder;
import com.supermarket.stockmanagement.model.Product;
import com.supermarket.stockmanagement.model.Supplier;
import com.supermarket.stockmanagement.model.Transaction;
import com.supermarket.stockmanagement.repository.PurchaseOrderRepository;
import com.supermarket.stockmanagement.repository.ProductRepository;
import com.supermarket.stockmanagement.repository.SupplierRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

@Service
public class PurchaseOrderService {

    @Autowired
    private PurchaseOrderRepository purchaseOrderRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private SupplierRepository supplierRepository;

    @Autowired
    private TransactionService transactionService;

    public List<PurchaseOrder> getAllPurchaseOrders() {
        return purchaseOrderRepository.findAll();
    }

    public PurchaseOrder getPurchaseOrderById(String id) {
        return purchaseOrderRepository.findById(id).orElse(null);
    }

    public List<PurchaseOrder> getPurchaseOrdersByStatus(PurchaseOrder.OrderStatus status) {
        return purchaseOrderRepository.findByStatus(status);
    }

    public List<PurchaseOrder> getPurchaseOrdersBySupplierId(String supplierId) {
        return purchaseOrderRepository.findBySupplierId(supplierId);
    }

    public List<PurchaseOrder> getPurchaseOrdersByProductId(String productId) {
        return purchaseOrderRepository.findByProductId(productId);
    }

    public List<PurchaseOrder> getUpcomingDeliveries() {
        LocalDateTime now = LocalDateTime.now();
        LocalDateTime weekLater = now.plusDays(7);
        return purchaseOrderRepository.findByExpectedDeliveryDateBetween(now, weekLater);
    }

    public PurchaseOrder createPurchaseOrder(PurchaseOrder purchaseOrder) {
        // Generate order number
        String orderNumber = generateOrderNumber();
        purchaseOrder.setOrderNumber(orderNumber);
        
        // Set order date
        purchaseOrder.setOrderDate(LocalDateTime.now());
        
        // Fetch product and supplier details
        Product product = productRepository.findById(purchaseOrder.getProductId()).orElse(null);
        Supplier supplier = supplierRepository.findById(purchaseOrder.getSupplierId()).orElse(null);
        
        if (product != null) {
            purchaseOrder.setProductName(product.getName());
        }
        
        if (supplier != null) {
            purchaseOrder.setSupplierName(supplier.getName());
        }
        
        // Calculate total amount
        if (purchaseOrder.getQuantity() != null && purchaseOrder.getUnitPrice() != null) {
            purchaseOrder.setTotalAmount(purchaseOrder.getQuantity() * purchaseOrder.getUnitPrice());
        }
        
        // Set default status
        if (purchaseOrder.getStatus() == null) {
            purchaseOrder.setStatus(PurchaseOrder.OrderStatus.PENDING);
        }
        
        if (purchaseOrder.getPaymentStatus() == null) {
            purchaseOrder.setPaymentStatus(PurchaseOrder.PaymentStatus.PENDING);
        }
        
        return purchaseOrderRepository.save(purchaseOrder);
    }

    public PurchaseOrder updatePurchaseOrder(String id, PurchaseOrder purchaseOrder) {
        PurchaseOrder existingOrder = purchaseOrderRepository.findById(id).orElse(null);
        if (existingOrder != null) {
            purchaseOrder.setId(id);
            purchaseOrder.setOrderNumber(existingOrder.getOrderNumber());
            purchaseOrder.setOrderDate(existingOrder.getOrderDate());
            
            // If status changes to DELIVERED, create a purchase transaction
            if (existingOrder.getStatus() != PurchaseOrder.OrderStatus.DELIVERED 
                && purchaseOrder.getStatus() == PurchaseOrder.OrderStatus.DELIVERED) {
                createPurchaseTransaction(purchaseOrder);
            }
            
            return purchaseOrderRepository.save(purchaseOrder);
        }
        return null;
    }

    public void deletePurchaseOrder(String id) {
        purchaseOrderRepository.deleteById(id);
    }

    private String generateOrderNumber() {
        LocalDateTime now = LocalDateTime.now();
        String dateStr = now.format(DateTimeFormatter.ofPattern("yyyyMMdd"));
        
        // Count orders for today
        LocalDateTime startOfDay = now.toLocalDate().atStartOfDay();
        LocalDateTime endOfDay = startOfDay.plusDays(1);
        long count = purchaseOrderRepository.countByOrderDateBetween(startOfDay, endOfDay);
        
        return String.format("PO-%s-%03d", dateStr, count + 1);
    }

    private void createPurchaseTransaction(PurchaseOrder order) {
        Transaction transaction = new Transaction();
        transaction.setProductId(order.getProductId());
        transaction.setType(Transaction.TransactionType.PURCHASE);
        transaction.setQuantity(order.getQuantity());
        transaction.setUnitPrice(order.getUnitPrice());
        transaction.setTotalAmount(order.getTotalAmount());
        transaction.setTransactionDate(LocalDateTime.now());
        transaction.setNotes("Purchase Order: " + order.getOrderNumber());
        
        transactionService.createTransaction(transaction);
    }

    public List<PurchaseOrder> getPendingPayments() {
        return purchaseOrderRepository.findByPaymentStatus(PurchaseOrder.PaymentStatus.PENDING);
    }

    public PurchaseOrder updatePaymentStatus(String id, PurchaseOrder.PaymentStatus status, String paymentMethod) {
        PurchaseOrder order = purchaseOrderRepository.findById(id).orElse(null);
        if (order != null) {
            order.setPaymentStatus(status);
            order.setPaymentMethod(paymentMethod);
            if (status == PurchaseOrder.PaymentStatus.PAID) {
                order.setPaymentDate(LocalDateTime.now());
            }
            return purchaseOrderRepository.save(order);
        }
        return null;
    }
}

File: frontend/src/app/services/purchase-order.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { PurchaseOrder, PaymentUpdate } from '../models/purchase-order.model';

@Injectable({
  providedIn: 'root'
})
export class PurchaseOrderService {
  private apiUrl = 'http://localhost:8080/api/purchase-orders';

  constructor(private http: HttpClient) { }

  getAllPurchaseOrders(): Observable<PurchaseOrder[]> {
    return this.http.get<PurchaseOrder[]>(this.apiUrl);
  }

  getPurchaseOrderById(id: string): Observable<PurchaseOrder> {
    return this.http.get<PurchaseOrder>(`${this.apiUrl}/${id}`);
  }

  getPurchaseOrdersByStatus(status: string): Observable<PurchaseOrder[]> {
    return this.http.get<PurchaseOrder[]>(`${this.apiUrl}/status/${status}`);
  }

  getPurchaseOrdersBySupplierId(supplierId: string): Observable<PurchaseOrder[]> {
    return this.http.get<PurchaseOrder[]>(`${this.apiUrl}/supplier/${supplierId}`);
  }

  getPurchaseOrdersByProductId(productId: string): Observable<PurchaseOrder[]> {
    return this.http.get<PurchaseOrder[]>(`${this.apiUrl}/product/${productId}`);
  }

  getUpcomingDeliveries(): Observable<PurchaseOrder[]> {
    return this.http.get<PurchaseOrder[]>(`${this.apiUrl}/upcoming-deliveries`);
  }

  getPendingPayments(): Observable<PurchaseOrder[]> {
    return this.http.get<PurchaseOrder[]>(`${this.apiUrl}/pending-payments`);
  }

  createPurchaseOrder(order: PurchaseOrder): Observable<PurchaseOrder> {
    return this.http.post<PurchaseOrder>(this.apiUrl, order);
  }

  updatePurchaseOrder(id: string, order: PurchaseOrder): Observable<PurchaseOrder> {
    return this.http.put<PurchaseOrder>(`${this.apiUrl}/${id}`, order);
  }

  updatePaymentStatus(id: string, paymentData: PaymentUpdate): Observable<PurchaseOrder> {
    return this.http.put<PurchaseOrder>(`${this.apiUrl}/${id}/payment`, paymentData);
  }

  deletePurchaseOrder(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
